modules/apps/portal-vulcan/portal-vulcan-impl/src/main/java/com/liferay/portal/vulcan/internal/dto/converter/DTOConverterRegistryImpl.java

```diff
@@ -35,33 +44,143 @@ public class DTOConverterRegistryImpl implements DTOConverterRegistry {
 
 	@Override
 	public Set<String> getDTOClassNames() {
-		return _serviceTrackerMap.keySet();
+		return _dtoConvertersMap.keySet();
 	}
 
 	@Override
 	public DTOConverter<?, ?> getDTOConverter(String dtoClassName) {
-		return _serviceTrackerMap.getService(dtoClassName);
+		List<DTOConverter<?, ?>> dtoConverters = _dtoConvertersMap.get(
+			dtoClassName);
+
+		if (ListUtil.isEmpty(dtoConverters)) {
+			return null;
+		}
+
+		return dtoConverters.get(0);
+	}
+
+	@Override
+	public DTOConverter<?, ?> getDTOConverter(
+		String applicationName, String dtoClassName, String version) {
+
+		List<DTOConverter<?, ?>> dtoConverters = _dtoConvertersMap.get(
+			dtoClassName);
+
+		if (ListUtil.isEmpty(dtoConverters)) {
+			return null;
+		}
+
+		for (DTOConverter<?, ?> dtoConverter : dtoConverters) {
+			Map<String, Object> properties = _dtoConverterPropertiesMap.get(
+				dtoConverter);
+
+			if (StringUtil.equals(
+					(String)properties.get("applicationName"),
+					applicationName) &&
+				StringUtil.equals((String)properties.get("version"), version)) {
+
+				return dtoConverter;
+			}
+		}
+
+		return null;
 	}
 
 	@Activate
 	protected void activate(BundleContext bundleContext) {
-		_serviceTrackerMap = ServiceTrackerMapFactory.openSingleValueMap(
-			bundleContext,
-			(Class<DTOConverter<?, ?>>)(Class<?>)DTOConverter.class,
-			"(dto.class.name=*)",
-			(serviceReference, emitter) -> {
-				String dtoClassName = (String)serviceReference.getProperty(
-					"dto.class.name");
-
-				emitter.emit(dtoClassName);
-			});
+		String filter = StringBundler.concat(
+			"(&(objectClass=", DTOConverter.class.getName(),
+			")(dto.class.name=*))");
+
+		_serviceTracker = ServiceTrackerFactory.open(
+			bundleContext, filter,
+			new DTOConverterServiceTrackerCustomizer(bundleContext));
 	}
 
 	@Deactivate
 	protected void deactivate() {
-		_serviceTrackerMap.close();
+		_serviceTracker.close();
 	}
 
-	private ServiceTrackerMap<String, DTOConverter<?, ?>> _serviceTrackerMap;
+	private final Map<DTOConverter<?, ?>, Map<String, Object>>
+		_dtoConverterPropertiesMap = new HashMap<>();
+	private final Map<String, List<DTOConverter<?, ?>>> _dtoConvertersMap =
+		new HashMap<>();
+	private ServiceTracker<DTOConverter<?, ?>, DTOConverter<?, ?>>
+		_serviceTracker;
+
+	private class DTOConverterServiceTrackerCustomizer
+		implements ServiceTrackerCustomizer
+			<DTOConverter<?, ?>, DTOConverter<?, ?>> {
+
+		@Override
+		public DTOConverter<?, ?> addingService(
+			ServiceReference<DTOConverter<?, ?>> serviceReference) {
+
+			String dtoClassName = (String)serviceReference.getProperty(
+				"dto.class.name");
+
+			DTOConverter<?, ?> dtoConverter = _bundleContext.getService(
+				serviceReference);
+
+			_dtoConvertersMap.compute(
```

Comment:
I've never seen `.compute` before (or I don't remember 😅 )
![](https://media.giphy.com/media/0FHJ363Octui8Emuul/giphy.gif)

Reply:
There are some interesing methods in Maps, but be careful, some of them are re-implemented in HashMap and the behaviour differs from the one implemented in Map